wrappers/common/common.go

```diff
@@ -42,14 +42,28 @@ func NewResponseWriter(w http.ResponseWriter) *ResponseWriter {
 	return &rw
 }
 
+func getParsersForRequest(ctx context.Context, r *http.Request) []trace.HTTPPropagator {
+	if trace.GlobalConfig.TraceHTTPHeaderParserHook != nil {
+		return trace.GlobalConfig.TraceHTTPHeaderParserHook(ctx, r)
+	}
+	return []trace.HTTPPropagator{
+		propagation.HoneycombHTTPPropagator{},
+	}
+}
+
 func StartSpanOrTraceFromHTTP(r *http.Request) (context.Context, *trace.Span) {
 	ctx := r.Context()
 	span := trace.GetSpanFromContext(ctx)
 	if span == nil {
-		// there is no trace yet. We should make one! and use the root span.
-		beelineHeader := r.Header.Get(propagation.TracePropagationHTTPHeader)
+		var spanContext *trace.SpanContext
+		for _, propagator := range getParsersForRequest(ctx, r) {
+			spanContext = propagator.Parse(ctx, r.Header)
+			spanContext = spanContext.MergeSpanContext(
+				trace.GetRemoteSpanContextFromContext(ctx),
```

Comment:
This behavior feels like a trap to me. I think we should accept all data from a single source and use that. If we want to examine multiple sources, that seems fine, but combining data from those sources feels like it will put us in a dangerous position. I would prefer to see us be clear about the order of importance if multiple headers are present and have a "most important wins" strategy.

Is there a user story style description of the circumstances in which merging data from multiple headers leads to the desired outcome?

Reply:
Clearest boundary:

- Multiple incoming headers, multiple propagators defined. 
- Last writer wins. Last defined propagator that has header(s) present is what will be used.
- Document the crap out of this in the hook definition.

Use case: 

- A service sits behind an ALB and that is how it is accessed from the public internet. That service can also be called directly from peer services within the same VPC. In this case, if a Honeycomb header exists, and NO amazon header exists, we should use it. If both exist, the Amazon header should clearly win.